### ISSUE TITLE
Yolov8 Tracking id is not assigned to the detected objects

### ISSUE THREAD

**Issue by VM8198** [ultralytics/ultralytics#6080]
### Search before asking

- [x] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Other

### Bug

I have trained a custom class to detect vehicle's number plate with Ultralytics documentation. The model works fine but the issue is  model.track() with Persist=True is not assigning Id's to the detected object. P.S. I'm getting the video stream via RTSP stream.

Image for reference
![github-issue](https://github.com/ultralytics/ultralytics/assets/26054305/61567d83-8fb9-4f34-8ca3-239d3f60a986)

### Environment

Ultralytics YOLOv8.0.202 🚀 Python-3.10.12 torch-2.1.0+cu121 CPU (Intel Core(TM) i7-7700 3.60GHz)
Setup complete ✅ (8 CPUs, 15.5 GB RAM, 52.8/456.9 GB disk)

OS                  Linux-6.2.0-35-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.12
Install                pip
RAM                 15.50 GB
CPU                 Intel Core(TM) i7-7700 3.60GHz
CUDA                None

matplotlib          ✅ 3.7.2>=3.3.0
numpy               ✅ 1.25.2>=1.22.2
opencv-python       ✅ 4.8.0.76>=4.6.0
pillow              ✅ 10.0.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.11.2>=1.4.1
torch               ✅ 2.1.0>=1.8.0
torchvision         ✅ 0.16.0>=0.9.0
tqdm                ✅ 4.66.1>=4.64.0
pandas              ✅ 2.0.3>=1.1.4
seaborn             ✅ 0.12.2>=0.11.0
psutil              ✅ 5.9.5
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1

### Minimal Reproducible Example

```
import cv2
from ultralytics import YOLO

model = YOLO('best.pt')

video_path = "rtsp stream url"
cap = cv2.VideoCapture(video_path)

while cap.isOpened():
    # Read a frame from the video
    success, frame = cap.read()

    if success:
        results = model.track(frame, persist=True)

        # Visualize the results on the frame
        annotated_frame = results[0].plot()

        # Display the annotated frame
        cv2.imshow("Frame", annotated_frame)
```

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by VM8198** [comment]
Any updates here? @RizwanMunawar @glenn-jocher

**Comment by glenn-jocher** [comment]
@VM8198 hello! 

We understand that you're keen on updates regarding the tracking ID assignment for detected objects in YOLOv8. Our team is actively investigating this and we aim to deliver a solution as soon as possible. 

Please bear with us while we work on this. Your patience and understanding are greatly appreciated. 

Thank you for using YOLOv8 and for contributing to our community with your question. We value your input in helping us improve our offerings.

Best,
Glenn Jocher and the Ultralytics Team

**Comment by VM8198** [comment]
Thanks for your response @glenn-jocher. I'm waiting for the updates!

**Comment by glenn-jocher** [comment]
You're welcome @VM8198! We appreciate your patience as we work on your issue with object ID tracking in YOLOv8. We understand this is an important feature for you and we're doing our best to address it. We'll ensure to keep you updated on our progress. Thanks for using YOLOv8 and for helping us improve with your valuable feedback!

**Comment by janbertoo** [comment]
Dear all,

I am running into the same issue. I have tried it with both bytetracker and botsort, but it won't output the object IDs. So at the moment the model.track() runs like model.predict().

Looking forward to the fix :)

**Comment by VM8198** [comment]
Hello @glenn-jocher, Any updates on this?

**Comment by glenn-jocher** [comment]
Hello @VM8198,

We're actively working on the issue. No updates just yet, but we'll inform you as soon as we have a resolution. Thank you for your patience.

**Comment by janbertoo** [comment]
Please do not close the issue, we are still waiting on a fix.

Thank you

**Comment by levismiller** [comment]
I'm seeing the same thing. Custom trained model which is model.track and model.persist behave the same. 
Running in Python: 
```
from ultralytics import YOLO
model_file = "/Users/bmo/yolo/football/runs/detect/train/weights/best.pt"
model = YOLO(model_file)
results = model.track(source=0, show=True, tracker="bytetrack.yaml", persist=True)
```
Mac OS: Sonoma 14.1.1

pip freeze 
certifi==2024.2.2
charset-normalizer==3.3.2
contourpy==1.2.0
cycler==0.12.1
Cython==3.0.8
filelock==3.13.1
fonttools==4.49.0
fsspec==2024.2.0
idna==3.6
Jinja2==3.1.3
kiwisolver==1.4.5
lapx==0.5.5
MarkupSafe==2.1.5
matplotlib==3.8.3
mpmath==1.3.0
networkx==3.2.1
numpy==1.26.4
opencv-python==4.9.0.80
packaging==23.2
pafy==0.5.5
pandas==2.2.0
pillow==10.2.0
psutil==5.9.8
py-cpuinfo==9.0.0
pyparsing==3.1.1
python-dateutil==2.8.2
pytz==2024.1
PyYAML==6.0.1
requests==2.31.0
scipy==1.12.0
seaborn==0.13.2
six==1.16.0
sympy==1.12
thop==0.1.1.post2209072238
torch==2.2.0
torchvision==0.17.0
tqdm==4.66.2
typing_extensions==4.9.0
tzdata==2024.1
ultralytics==8.1.14
urllib3==2.2.0
youtube-dl==2020.12.2

**Comment by glenn-jocher** [comment]
@levismiller thank you for providing detailed information about your issue. It seems like you've done everything correctly based on the information provided. The `model.track()` function with `persist=True` should indeed maintain track IDs across frames if the objects are consistently detected.

Here are a few steps you can take to troubleshoot the issue:

1. **Check Tracker Configuration:**
   Ensure that the tracker configuration is set up correctly. If you're using a custom tracker configuration, verify that it's properly defined and that the tracker type supports ID assignment.

2. **Verify Object Detection Consistency:**
   Tracking relies on consistent detection of objects across frames. If the detection model occasionally misses the number plates or if there's significant variation in detection confidence, the tracker might lose track of the object. You can check the detection results without tracking to ensure that the number plates are consistently detected.

3. **Debugging Information:**
   Add debugging information to your loop to print out the detection results and track IDs. This can help you understand if the IDs are being generated but not displayed, or if they're not being generated at all.

4. **Update to the Latest Version:**
   Ensure that you are using the latest version of the Ultralytics YOLOv8 repository. If not, update your repository and try again, as there may have been fixes or improvements related to tracking.

5. **Check the Tracker's Limitations:**
   Some trackers may have limitations on the number of objects they can track or may require certain conditions to be met for successful tracking. Review the documentation for the specific tracker you're using to ensure that your use case is supported.

6. **Review the Tracker's Output:**
   The tracker's output should include the `id` attribute for each detected object. If this attribute is missing or not being updated correctly, it could indicate an issue with the tracker itself.

7. **Submit an Issue:**
   If you've gone through the above steps and the problem persists, consider submitting an issue on the Ultralytics YOLOv8 GitHub repository with all the details you've gathered. The maintainers and community may be able to provide further assistance.

Here's an example of how you might add debugging information to your loop:

```python
while cap.isOpened():
    success, frame = cap.read()
    if success:
        results = model.track(frame, persist=True)
        annotated_frame = results[0].plot()

        # Debugging: Print track IDs
        if results[0].boxes.id is not None:
            print("Track IDs:", results[0].boxes.id)

        cv2.imshow("Frame", annotated_frame)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break
```

Remember to check the console output for the printed track IDs to see if they are being generated. If the IDs are printed but not shown in the annotated frame, there might be an issue with the `plot()` method or the visualization code. If the IDs are not printed, the issue likely lies with the tracking logic itself.

**Comment by pcismyname** [comment]
Still facing this issue. Please don't close it yet. I got this when I use model.track(source=src, show=False, stream=True, verbose=False, classes=classes, conf=conf, iou=iou, max_det=max_det, tracker=tracker_config, persist=True) and it occurs only in some input (difference video) the size of result is 6 (no tracking ID) instead of 7.

**Comment by glenn-jocher** [comment]
@pcismyname hello there! 👋

Thanks for keeping us updated on your issue. It seems like the inconsistency in track IDs across different video inputs might be related to the specific characteristics of those videos. However, one potential quick fix you could try involves ensuring that the `results` are being handled correctly for videos without track IDs. Here's a small snippet to help manage cases when the size of `result` differs:

```python
results = model.track(source=src, show=False, stream=True, verbose=False, classes=classes, conf=conf, iou=iou, max_det=max_det, tracker=tracker_config, persist=True)

# Check if tracking ID is in the results for debugging
if len(results[0].xyxy[0]) == 7:
    # Process normally
    print("Tracking IDs found.")
else:
    print("Tracking IDs missing in this frame.")

```

This might help in pinpointing when exactly the IDs are getting dropped. We're here to help you smooth out these wrinkles, so let's keep this issue open until we find a solid solution. Thanks for sticking with us! 🛠️

**Comment by VM8198** [comment]
@glenn-jocher I think the issue here is with tracker. Tracker is currently working with the pretrained YOLO classes only. If custom classes are used than any of the tracker seems not to be working. Is it so that we need to train the tracker with custom classes? Like DeepSORT needs to be trained for custom classes? what are your views on this?

**Comment by glenn-jocher** [comment]
@VM8198 hi there! 👋

Great question! Our trackers are designed to work out-of-the-box with both pretrained YOLO classes and custom classes without the need for retraining the tracker itself. However, for optimal performance with custom classes, particularly in complex or specific environments, fine-tuning the tracking algorithm might be beneficial. This could involve customizing the tracking configuration or integrating post-processing steps tailored to your use case.

Here's a quick tip to ensure your custom classes are correctly integrated:

```python
# Load your custom trained model
model = YOLO('path/to/your/custom/model.pt')

# Perform tracking on your video or stream
results = model.track(source='your_video.mp4', persist=True)
```

Make sure your `model.pt` was trained with the custom classes. The tracker should automatically use these classes for tracking without extra steps. However, specific tracking enhancements or adjustments are always a bonus!

Let me know if this helps or if you have more questions. Happy tracking! 🚀

**Comment by kurkurzz** [comment]
Hi still facing this issue. Any new updates on this? My initial expectation is every detected object should get an ID. why does some have and the rest did not?

**Comment by glenn-jocher** [comment]
Thank you for your patience. Please ensure you are using the latest version of the Ultralytics package. If the issue persists, it may be related to detection consistency or tracker configuration.
